src/plugins/matrixops/flip_operation.cpp

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2018 Bita Hasheminezhad
+// Copyright (c) 2018 Hartmut Kaiser
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/execution_tree/primitives/node_data_helpers.hpp>
+#include <phylanx/ir/node_data.hpp>
+#include <phylanx/plugins/matrixops/flip_operation.hpp>
+#include <phylanx/util/matrix_iterators.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+#include <hpx/util/optional.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <blaze/Math.h>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const flip_operation::match_data =
+    {
+        hpx::util::make_tuple("flip",
+        std::vector<std::string>{"flip(_1,_2)"},
+        &create_flip_operation, &create_primitive<flip_operation>,
+        "a, axis\n"
+        "Args:\n"
+        "\n"
+        "    a (vector or matrix) : a vector or a matrix\n"
+        "    axis (integer): an axis to flip along\n"
+        "\n"
+        "Returns:\n"
+        "\n"
+        "Reverses the order of elements in an array along the given axis")
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    flip_operation::flip_operation(primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    template <typename T>
+    primitive_argument_type flip_operation::flip1d(ir::node_data<T>&& arg,
+        val_type axis) const
+    {
+        if (axis != 0 && axis != -1)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "flip_operation::flip1d",
+                generate_error_message(
+                    "the flip_operation primitive requires operand axis to be "
+                    "either 0 or -1 for vectors."));
+        }
+
+        auto v = arg.vector();
+        blaze::DynamicVector<T> result(v.size());
+        std::reverse_copy(std::begin(v), std::end(v), std::begin(result));
+
```

Comment:
If `arg` is not a reference you should be able to avoid allocating the result. In this case the operation could be performed in-place. I.e. something like:
```
if (!arg.is_ref())
{
    auto v = arg.vector(); 
    std::reverse(std::begin(v), std::end(v));
    return primitive_argument_type{std::move(arg)};
}
else
{
    auto v = arg.vector(); 
    blaze::DynamicVector<T> result(v.size()); 
    std::reverse_copy(std::begin(v), std::end(v), std::begin(result));
    return primitive_argument_type{std::move(result)};
}
```
should do the trick.